File: InterviewTraining/js/solutions/linkedListCycle.js
/*
Linked List Cycle Detection

Write a function that determines if a linked list contains a cycle.
A cycle occurs when a node's next pointer points to a previous node in the list,
creating a loop.

What is a cycle?
- A cycle exists when you can reach the same node again by following 'next' pointers
- The cycle can occur anywhere in the linked list
- Not all nodes need to be part of the cycle

Examples:

1. List with cycle:
   1 -> 2 -> 3 -> 4 -> 5 -> 3 (points back to 3)
   Returns: true

2. List without cycle:
   1 -> 2 -> 3 -> 4 -> null
   Returns: false

3. Single node cycle:
   1 -> 1 (points to itself)
   Returns: true

4. Empty list:
   null
   Returns: false

Parameters:
- head: The first node of a linked list
  * Each node has properties:
    - value: The node's value
    - next: Reference to next node or null

Return:
- boolean: true if cycle exists, false otherwise

Hints:
- Think about using two pointers moving at different speeds
- What happens when two people run around a track at different speeds?
- Consider memory usage - can you solve it with O(1) space?
*/
function hasCycle(head) {
    let slow = head, fast = head;
    while (fast && fast.next) {
        slow = slow.next;
        fast = fast.next.next;
        if (slow === fast) return true;
    }
    return false;
}

module.exports = hasCycle;

File: InterviewTraining/go/solutions/linkedListCycle.go
//go:build problem

package main

// Solution accepts nodes JSON and headId and returns whether a cycle exists.
func Solution(nodes []map[string]interface{}, headId string) bool {
    // build next map
    next := map[string]*string{}
    for _, n := range nodes {
        id := n["id"].(string)
        if n["next"] == nil {
            next[id] = nil
        } else {
            s := n["next"].(string)
            next[id] = &s
        }
    }
    // tortoise and hare using ids
    if headId == "" {
        return false
    }
    slow := headId
    fast := headId
    for {
        // advance slow by 1
        if next[slow] == nil {
            return false
        }
        slow = *next[slow]
        // advance fast by 2
        if next[fast] == nil {
            return false
        }
        f1 := next[fast]
        if f1 == nil {
            return false
        }
        if next[*f1] == nil {
            return false
        }
        fast = *next[*f1]
        if slow == fast {
            return true
        }
    }
}
